test/unit/puppetlabs/services/jruby/jruby_puppet_core_test.clj

```diff
@@ -13,6 +14,37 @@
     :ruby-load-path ["./ruby/puppet/lib" "./ruby/facter/lib"]},
    :certificate-authority {:certificate-status {:client-whitelist []}}})
 
+(defmacro with-stdin-str
+  "Evaluates body in a context in which System/in is bound to a fresh
+  input stream initialized with the string s.  The return value of evaluating
+  body is returned."
+  [s & body]
+  `(let [system-input# (System/in)
+         string-input# (new ByteArrayInputStream (.getBytes ~s))]
+     (try
+       (System/setIn string-input#)
+       ~@body
+       (finally (System/setIn system-input#)))))
+
+(defmacro capture-out
+  "capture System.out and return it as the value of :out in the return map.
+  The return value of body is available as :return in the return map.
+
+  This macro is not thread safe."
+  [& body]
+  `(let [return-map# (atom {})
+         system-output# (System/out)
+         captured-output# (new ByteArrayOutputStream)
+         capturing-print-stream# (new PrintStream captured-output#)]
+     (try
+       (System/setOut capturing-print-stream#)
+       (swap! return-map# assoc :return (do ~@body))
+       (finally
+         (.flush capturing-print-stream#)
+         (swap! return-map# assoc :out (.toString captured-output#))
+         (System/setOut system-output#)))
+     @return-map#))
```

Comment:
Yeah, they make me nervous too.  The original implementation was using functions but was really awkward.  I'd like to use a binding in clojure, but my assumption is that won't help with `System.out` and `System.in`.

My main concerns are around thread safety, but I can't imagine anything else mucking with System.out unless we do something like [parallel-test](https://github.com/aredington/parallel-test) which requires tests to be marked as `^:parallel` before they'll run concurrently.

My thinking was this isn't really any worse than how we're doing other things like TCP 8140, but it's also far from ideal and has lots of room for improvement.